src/subscriptions/allowListUtil.ts

```diff
@@ -0,0 +1,82 @@
+import { groupBy } from 'lodash';
+import { makeLogger } from 'fhir-works-on-aws-interface';
+import getAllowListedSubscriptionEndpoints from './allowList';
+
+const SINGLE_TENANT_ALLOW_LIST_KEY = 'SINGLE_TENANT_ALLOW_LIST_KEY';
+const logger = makeLogger({ component: 'subscriptions' });
+
+export interface AllowListInfo {
+    allowList: (string | RegExp)[];
+    headerMap: { [key: string]: string[] };
+}
+
+const extractAllowListInfo = (subscriptionEndpoints: SubscriptionEndpoint[]): AllowListInfo => {
+    const allowList: (string | RegExp)[] = [];
+    const headerMap: { [key: string]: string[] } = {};
+    subscriptionEndpoints.forEach((allowEndpoint: SubscriptionEndpoint) => {
+        allowList.push(allowEndpoint.endpoint);
+        headerMap[allowEndpoint.endpoint.toString()] = allowEndpoint.headers || [];
+    });
+    return { allowList, headerMap };
+};
+
+export async function getAllowListInfo({
+    enableMultitenancy = false,
+}: {
+    enableMultitenancy: boolean;
+}): Promise<{ [key: string]: AllowListInfo }> {
+    const originalAllowList = await getAllowListedSubscriptionEndpoints();
+    logger.debug(originalAllowList);
+    if (!enableMultitenancy) {
+        return { [SINGLE_TENANT_ALLOW_LIST_KEY]: extractAllowListInfo(originalAllowList) };
+    }
+    const allowListInfo: { [key: string]: AllowListInfo } = {};
+    const endpointsGroupByTenant: { [key: string]: SubscriptionEndpoint[] } = groupBy(
+        originalAllowList,
+        (allowEndpoint: SubscriptionEndpoint) => allowEndpoint.tenantId,
+    );
+    Object.entries(endpointsGroupByTenant).forEach(([key, value]) => {
+        allowListInfo[key] = extractAllowListInfo(value);
+    });
+    return allowListInfo;
+}
+
+/**
+ * Verify endpoint is allow listed
+ * Return allow list headers if endpoint is allow listed
+ * Throw error if endpoint is not allow listed
+ * @param allowListInfo
+ * @param endpoint
+ * @param tenantId
+ * @param enableMultitenancy
+ */
+export const getAllowListHeaders = (
+    allowListInfo: { [key: string]: AllowListInfo },
+    endpoint: string,
+    { enableMultitenancy = false, tenantId }: { enableMultitenancy: boolean; tenantId: string | undefined },
+): string[] => {
+    const getHeaders = ({ allowList, headerMap }: AllowListInfo): string[] => {
+        // eslint-disable-next-line no-restricted-syntax
+        for (const allowedEndpoint of allowList) {
+            if (allowedEndpoint instanceof RegExp && allowedEndpoint.test(endpoint)) {
+                return headerMap[allowedEndpoint.toString()];
+            }
+            if (allowedEndpoint === endpoint) {
+                return headerMap[allowedEndpoint];
+            }
+        }
+        throw new Error(`Endpoint ${endpoint} is not allow listed.`);
+    };
+
+    if (enableMultitenancy) {
+        if (tenantId) {
+            return getHeaders(allowListInfo[tenantId]);
```

Comment:
what happens if there are no entries in the allow list for the given `tenantId`?

I'm doing some tests and I get an error:
```
"errorType": "TypeError",
    "errorMessage": "Cannot destructure property 'allowList' of 'undefined' as it is undefined.",
    "stack": [
        "TypeError: Cannot destructure property 'allowList' of 'undefined' as it is undefined.",
        "    at getHeaders (/var/task/subscriptions/webpack:/Users/nestorba/Amazon/AWS/github/workspace/fhir-works-on-aws-deployment/subscriptions/restHook.ts:58:27)",
        "    at getAllowListHeaders (/var/task/subscriptions/webpack:/Users/nestorba/Amazon/AWS/github/workspace/fhir-works-on-aws-deployment/subscriptions/restHook.ts:73:20)",
        "    at /var/task/subscriptions/webpack:/Users/nestorba/Amazon/AWS/github/workspace/fhir-works-on-aws-deployment/subscriptions/restHook.ts:114:34",
        "    at Array.map (<anonymous>)",
        "    at Runtime.sendRestHookNotification [as handler] (/var/task/subscriptions/webpack:/Users/nestorba/Amazon/AWS/github/workspace/fhir-works-on-aws-deployment/subscriptions/restHook.ts:109:48)"
    ]
```

Maybe my dev setup is messed up, but please double check if this is a bug

Reply:
Thanks! I'll check on it.